File: Android/VideoComposeSample/app/src/main/java/com/example/videosamplecompose/TrtcViewModel.kt
package com.example.videosamplecompose

import android.content.Context
import android.os.Bundle
import android.util.Log
import androidx.compose.runtime.mutableStateListOf
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import com.tencent.trtc.TRTCCloud
import com.tencent.trtc.TRTCCloudDef
import com.tencent.trtc.TRTCCloudListener

class RemoteUser(var userId: String) : ViewModel() {
    var videoAvailable = MutableLiveData(false)
    var audioAvailable = MutableLiveData(false)
}

class TrtcViewModel(var debugPreview: Boolean = false): ViewModel()  {
    val joined = MutableLiveData(false)
    val errorCode = MutableLiveData(0)
    val errorMessage = MutableLiveData("")
    val userId = MutableLiveData("")
    val roomId = MutableLiveData(0)
    val remoteUsers = MutableLiveData<MutableList<RemoteUser>>(mutableStateListOf())
    val audioAvailable = MutableLiveData(false)
    val videoAvailable = MutableLiveData(false)
    val isFrontCamera = MutableLiveData(true)
    private lateinit var trtcCloud: TRTCCloud

    init {
        if (debugPreview) {
            userId.value = "User1"
            for (i in 1..3) {
                val id = "user$i"
                val user = RemoteUser(id)
                remoteUsers.value?.add(user)
            }
        }
    }

    fun setup(context: Context) {
        trtcCloud = TRTCCloud.sharedInstance(context)
        trtcCloud.setListener(TRTCCloudImplListener())

        userId.value = "User-" + genRandomString(4)
        videoAvailable.value = true
        audioAvailable.value = true
    }

    fun join(roomId: Int) {
        this.roomId.value = roomId

        // workaround: stop the video once
        videoAvailable.value = false
        trtcCloud.stopLocalPreview()

        val sdkSecret = TrtcUserSig()
        val trtcParams = TRTCCloudDef.TRTCParams()
        trtcParams.sdkAppId = sdkSecret.SDKAPPID
        trtcParams.userId = userId.value
        trtcParams.roomId = roomId
        trtcParams.userSig = sdkSecret.genTestUserSig(trtcParams.userId)
        //trtcParams.startLocalPreview(isFrontCamera, view)
        trtcCloud.startLocalAudio(TRTCCloudDef.TRTC_AUDIO_QUALITY_SPEECH)
        trtcCloud.enterRoom(trtcParams, TRTCCloudDef.TRTC_APP_SCENE_VIDEOCALL)
    }

    fun leave() {
        trtcCloud.stopLocalAudio()
        //trtcCloud.stopLocalPreview()
        trtcCloud.exitRoom()
    }

    fun muteLocalVideo(mute: Boolean) {
        trtcCloud.muteLocalVideo(TRTCCloudDef.TRTC_VIDEO_STREAM_TYPE_BIG, mute)
        videoAvailable.value = !mute
    }

    fun muteLocalAudio(mute: Boolean) {
        trtcCloud.muteLocalAudio(mute)
        audioAvailable.value = !mute
    }

    fun switchCamera(isFrontCamera: Boolean? = null) {
        if (isFrontCamera == null) {
            this.isFrontCamera.value = !this.isFrontCamera.value!!
        } else {
            this.isFrontCamera.value = isFrontCamera
        }
        trtcCloud.deviceManager.switchCamera(this.isFrontCamera.value!!)
    }

    private inner class TRTCCloudImplListener : TRTCCloudListener() {
        private val logTag = "[MyDebug:TRTCCloudListener]"

        override fun onEnterRoom(result: Long) {
            Log.d(logTag, "onEnterRoom result: $result")
            joined.value = true
            videoAvailable.value = true
        }

        override fun onExitRoom(reason: Int) {
            Log.d(logTag, "onExitRoom reason: $reason")
            joined.value = false
        }

        override fun onRemoteUserEnterRoom(userId: String?) {
            Log.d(logTag, "onRemoteUserEnterRoom userId: $userId")
            if (userId != null) {
                val user = RemoteUser(userId)
                remoteUsers.value?.add(user)
            }
        }

        override fun onRemoteUserLeaveRoom(userId: String?, reason: Int) {
            Log.d(logTag, "onRemoteUserLeaveRoom userId: $userId, reason: $reason")
            val user = findRemoteUser(userId)
            if (user != null) {
                remoteUsers.value?.remove(user)
            }
        }

        override fun onUserAudioAvailable(userId: String?, available: Boolean) {
            Log.d(logTag, "onUserAudioAvailable userId: $userId, available: $available")
            val user = findRemoteUser(userId)
            if (user != null) {
                user.audioAvailable.value = available
            }
        }

        override fun onUserVideoAvailable(userId: String, available: Boolean) {
            Log.d(logTag, "onUserVideoAvailable userId: $userId, available: $available")
            val user = findRemoteUser(userId)
            if (user != null) {
                user.videoAvailable.value = available
            }
        }

        override fun onError(errCode: Int, errMsg: String, extraInfo: Bundle) {
            Log.d(logTag, "onError errCode: $errCode, errMsg: $errMsg")
        }
    }

    private fun findRemoteUser(userId: String?): RemoteUser? {
        if (userId != null) {
            val users = remoteUsers.value?.filter { user ->
                user.userId == userId
            }
            if (users != null && users.isNotEmpty()) {
                return users[0]
            }
        }
        return null
    }

    private fun genRandomString(length: Int) : String {
        val charset = "ABCDEFGHIJKLMNOPQRSTUVWXTZabcdefghiklmnopqrstuvwxyz0123456789"
        return (1..length)
            .map { charset.random() }
            .joinToString("")
    }
}

File: iOS/VideoSwiftUISample/VideoSwiftUISample/TrtcViewModel.swift
//
//  RemoteUser.swift
//  VideoSwiftUISample
//

import Foundation
import TXLiteAVSDK_TRTC

class RemoteUser: Identifiable, ObservableObject {
    var id: UUID
    var userId: String
    @Published var videoAvailable: Bool
    @Published var audioAvailable: Bool
    
    init(id: UUID = UUID(), userId: String, videoAvailable: Bool = false, audioAvailable: Bool = false) {
        self.id = id
        self.userId = userId
        self.videoAvailable = videoAvailable
        self.audioAvailable = audioAvailable
    }
}

class TrtcViewModel: NSObject, ObservableObject {
    @Published var remoteUsers: [RemoteUser] = []
    @Published var errCode: Int32 = 0
    @Published var errMsg = ""
    @Published var joined = false
    @Published var userId : String = ""
    @Published var roomId : Int = 1
    @Published var audioAvailable = true
    @Published var videoAvailable = true
    @Published var isFrontCamera = true
    
    private var trtcCloud: TRTCCloud = TRTCCloud.sharedInstance()
    private var maxRemoteUserCount = 3
    
    init(debugPreview: Bool = false) {
        super.init()
        
        trtcCloud.delegate = self
        userId = "User-" + generator(4)
        
        if debugPreview {
            for i in 1..<4 {
                let user = RemoteUser(userId: "user\(i)")
                remoteUsers.append(user)
            }
        }
    }
    
    func enterRoom() {
        errCode = 0
        errMsg = ""
        
        // stop the video once
        videoAvailable = false
        trtcCloud.stopLocalPreview()
        
        let params = TRTCParams()
        params.sdkAppId = UInt32(SDKAPPID)
        params.roomId = UInt32(roomId)
        params.userId = userId
        params.role = .anchor
        params.userSig = TrtcUserSig.genTestUserSig(identifier: userId) as String
        trtcCloud.enterRoom(params, appScene: .videoCall)
        trtcCloud.startLocalAudio(.music)
    }
    
    func exitRoom() {
        trtcCloud.stopLocalAudio()
        //trtcCloud.stopLocalPreview()
        trtcCloud.exitRoom()
    }
    
    func muteLocalAudio(mute: Bool) {
        trtcCloud.muteLocalAudio(mute)
        audioAvailable = !mute
    }
    
    func muteLocalVideo(mute: Bool) {
        trtcCloud.muteLocalVideo(.big, mute: mute)
        videoAvailable = !mute
    }
    
    func switchCamera(_ isFrontCamera: Bool? = nil) {
        if let isFrontCamera = isFrontCamera {
            self.isFrontCamera = isFrontCamera
        } else {
            self.isFrontCamera.toggle()
        }
        trtcCloud.getDeviceManager().switchCamera(self.isFrontCamera)
    }

    private func generator(_ length: Int) -> String {
        let letters = "ABCDEFGHIJKLMNOPQRSTUVWXTZabcdefghiklmnopqrstuvwxyz0123456789"
        var randomString = ""
        for _ in 0 ..< length {
            randomString += String(letters.randomElement()!)
        }
        return randomString
    }
}

extension TrtcViewModel: TRTCCloudDelegate {
    func onEnterRoom(_ result: Int) {
        print("*** onEnterRoom: result: \(result)")
        joined = true
        videoAvailable = true
    }
    
    func onExitRoom(_ reason: Int) {
        print("*** onExitRoom: reason: \(reason)")
        joined = false
        remoteUsers.removeAll()
    }
    
    func onRemoteUserEnterRoom(_ userId: String) {
        print("*** onRemoteUserEnterRoom: userId: \(userId)")
        if remoteUsers.count >= maxRemoteUserCount {
            return // TODO: Allow flexibility in number of participants
        }
        let remoteUser = RemoteUser(userId: userId)
        remoteUsers.append(remoteUser)
    }

    func onRemoteUserLeaveRoom(_ userId: String, reason: Int) {
        print("*** onRemoteUserLeaveRoom: userId: \(userId), reason: \(reason)")
        remoteUsers.removeAll(where: { $0.userId == userId })
    }
    
    func onUserAudioAvailable(_ userId: String, available: Bool) {
        print("*** onUserAudioAvailable: userId: \(userId): available: \(available)")
        if let remoteUser = remoteUsers.first(where: { $0.userId == userId }) {
            remoteUser.audioAvailable = available
        }
    }
    
    func onUserVideoAvailable(_ userId: String, available: Bool) {
        print("*** onUserVideoAvailable: userId: \(userId): available: \(available)")
        if let remoteUser = remoteUsers.first(where: { $0.userId == userId }) {
            remoteUser.videoAvailable = available
        }
    }
    
    func onError(_ errCode: TXLiteAVError, errMsg: String?, extInfo: [AnyHashable : Any]?) {
        self.errCode = errCode.rawValue
        self.errMsg = ""
        if let errMsg = errMsg {
            self.errMsg = errMsg
        }
        print("*** onError: \(self.errCode): \(self.errMsg)")
    }
}
